diff --git a/src/sundials/CMakeLists.txt b/src/sundials/CMakeLists.txt
index c7f1f90ee..02399c97b 100644
--- a/src/sundials/CMakeLists.txt
+++ b/src/sundials/CMakeLists.txt
@@ -113,9 +113,12 @@ endif()
 add_prefix(${SUNDIALS_SOURCE_DIR}/src/sundials/ sundials_SOURCES)
 
 if(ENABLE_MPI)
-  set(_link_mpi_if_needed PUBLIC MPI::MPI_C)
+  set(_link_mpi_if_needed PUBLIC 
+    MPI::MPI_C
+    $<$<LINK_LANGUAGE:CXX>:MPI::MPI_CXX>)
 endif()
 
+
 if(SUNDIALS_BUILD_WITH_PROFILING)
   if(ENABLE_CALIPER)
     set(_link_caliper_if_needed PUBLIC caliper)
