--- a/src/PYB11/CMakeLists.txt	2019-12-17 13:38:56.342201000 -0800
+++ b/src/PYB11/CMakeLists.txt	2019-12-17 13:25:01.707132000 -0800
@@ -56,6 +56,7 @@
 include_directories("${PYTHON_INCLUDE_DIR}")
 include_directories("${PYB11_DIR}")
 include_directories("${CMAKE_INSTALL_PREFIX}/include")
+include_directories("${PYBIND11_INCLUDE_DIRS}")
 
 set(PYTHON_LINK "python${PYTHON_VERSION}")
 
@@ -85,7 +86,8 @@
 set_target_properties(pypolytope PROPERTIES OUTPUT_NAME polytope)
 target_link_libraries(pypolytope
                       #${PYTHON_LINK}
-                      ${POLYTOPE_STATIC_LIB}
+                      #${POLYTOPE_STATIC_LIB}
+                      polytope
                       ${TRIANGLE_STATIC_LIB}
                       ${TETGEN_STATIC_LIB})
 
