diff --git a/src/thirdparty/axom/fmt/core.h b/src/thirdparty/axom/fmt/core.h
index dbfe1cd86..9393a8710 100644
--- a/src/thirdparty/axom/fmt/core.h
+++ b/src/thirdparty/axom/fmt/core.h
@@ -106,13 +106,14 @@
 #  define AXOM_FMT_CONSTEXPR
 #endif
 
-#if ((AXOM_FMT_CPLUSPLUS >= 202002L) &&                            \
-     (!defined(_GLIBCXX_RELEASE) || _GLIBCXX_RELEASE > 9)) || \
-    (AXOM_FMT_CPLUSPLUS >= 201709L && AXOM_FMT_GCC_VERSION >= 1002)
-#  define AXOM_FMT_CONSTEXPR20 constexpr
-#else
-#  define AXOM_FMT_CONSTEXPR20
-#endif
+// #if ((AXOM_FMT_CPLUSPLUS >= 202002L) &&                            \
+//      (!defined(_GLIBCXX_RELEASE) || _GLIBCXX_RELEASE > 9)) || \
+//     (AXOM_FMT_CPLUSPLUS >= 201709L && AXOM_FMT_GCC_VERSION >= 1002)
+// #  define AXOM_FMT_CONSTEXPR20 constexpr
+// #else
+// #  define AXOM_FMT_CONSTEXPR20
+// #endif
+#define AXOM_FMT_CONSTEXPR20
 
 // Check if constexpr std::char_traits<>::{compare,length} are supported.
 #if defined(__GLIBCXX__)
